runtime/drivers/duckdb/connectors.go

```diff
@@ -123,14 +156,158 @@ func (c *connection) ingestLocalFiles(ctx context.Context, env *connectors.Env,
+		from, err = sourceReader(files, format, addSchemaInference(a.ingestionProps, srcSchema))
+		if err != nil {
+			return err
+		}
+	}
+
+	colNames := strings.Join(keys(srcSchema), ",")
+	query = fmt.Sprintf("INSERT INTO %q (%s) (SELECT %s FROM %s);", a.source.Name, colNames, colNames, from)
+	return a.Exec(ctx, &drivers.Statement{Query: query, Priority: 1})
+}
+
+// updateSchema updates the schema of the table in case new file adds a new column or
+// updates the datatypes of an existing columns with a wider datatype.
+func (a *appender) updateSchema(ctx context.Context, from string, fileNames []string) (srcSchema map[string]string, err error) {
+	// schema of new files
+	if srcSchema, err = a.scanSchemaFromQuery(ctx, fmt.Sprintf("DESCRIBE (SELECT * FROM %s LIMIT 0);", from)); err != nil {
+		return
+	}
+
+	// combined schema
+	qry := fmt.Sprintf("DESCRIBE ((SELECT * FROM %s limit 0) UNION ALL BY NAME (SELECT * FROM %s limit 0));", a.source.Name, from)
+	unionSchema, err := a.scanSchemaFromQuery(ctx, qry)
+	if err != nil {
+		return nil, err
+	}
+
+	// current schema
+	if a.tableSchema == nil {
+		a.tableSchema, err = a.scanSchemaFromQuery(ctx, fmt.Sprintf("DESCRIBE %s;", a.source.Name))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	newCols := make(map[string]string)
+	colTypeChanged := make(map[string]string)
+	for colName, colType := range unionSchema {
+		oldType, ok := a.tableSchema[colName]
+		if !ok {
+			newCols[colName] = colType
+		} else if oldType != colType {
+			colTypeChanged[colName] = colType
+		}
+	}
+
+	if !a.allowColRelaxation {
+		if len(srcSchema) < len(unionSchema) {
+			fileNames := strings.Join(names(fileNames), ",")
+			columns := strings.Join(missingMapKeys(a.tableSchema, srcSchema), ",")
+			a.logger.Error("new files are missing columns and column relaxation not allowed", zap.String("files", fileNames),
+				zap.String("columns", columns))
+			return nil, fmt.Errorf("new files %q are missing columns %q and schema relaxation not allowed", fileNames, columns)
+		}
+
+		if len(colTypeChanged) != 0 {
+			fileNames := strings.Join(names(fileNames), ",")
+			columns := strings.Join(keys(colTypeChanged), ",")
+			a.logger.Error("new files change datatypes of some columns and column relaxation not allowed", zap.String("files", fileNames),
+				zap.String("columns", columns))
+			return nil, fmt.Errorf("new files %q change datatypes of some columns %q and column relaxation not allowed", fileNames, columns)
+		}
+	}
+
+	if len(newCols) != 0 && !a.allowColAddition {
+		fileNames := strings.Join(names(fileNames), ",")
+		columns := strings.Join(missingMapKeys(srcSchema, a.tableSchema), ",")
+		a.logger.Error("new files have new columns and column addition not allowed", zap.String("files", fileNames),
+			zap.String("columns", columns))
+		return nil, fmt.Errorf("new files %q have new columns %q and column addition not allowed", fileNames, columns)
+	}
```

Comment:
Can remove duplicate log statements, and just log any error returned from `updateSchema` when calling it in `appendData` (unless there's something even higher in the call chain that already logs errors?)